src/webgpu/shader/types.ts

```diff
@@ -212,33 +261,43 @@ export function* generateTypes({
 
   // Array types
   if (containerType === 'array') {
+    // Buffer affective binding size must be a multiple of 4. Adjust array length as needed.
+    let arrayLength = kDefaultArrayLength;
+    if (
+      addressSpace === 'storage' &&
+      scalarInfo.layout !== undefined &&
+      scalarInfo.layout.alignment % 4 !== 0
+    ) {
+      arrayLength = align(arrayLength, 4);
+    }
+
     const arrayTypeInfo = {
       elementBaseType: `${baseType}`,
-      arrayLength: kArrayLength,
+      arrayLength,
       layout: scalarInfo.layout
         ? {
             alignment: scalarInfo.layout.alignment,
```

Comment:
aside: I'm not exactly sure what the type layouts are supposed to express, but should this be 16 in uniform storage class?

(I did not change this in the code above.)